File: 0001_simple_db_in_memory_index/python/test_simple_db_in_memory_index.py
import os
import pytest
import simple_db_in_memory_index


@pytest.fixture
async def filename():
    filename = 'test.db'
    yield filename
    os.remove(filename)


@pytest.fixture
async def db(filename):
    db = simple_db_in_memory_index.SimpleDbInMemoryIndex(filename=filename)
    await db.set('greeting',  {'hello': 'world'})
    await db.set('menu',  {
        'breakfast': 'bubur ayam', 'lunch': 'nasi rendang', 'dinner': 'nasi goreng'
    })
    yield db


async def test_existing_db(db):
    db2 = simple_db_in_memory_index.SimpleDbInMemoryIndex(filename=db.filename)
    assert db2._index._idx_map == {'greeting': (0, 28), 'menu': (28, 83)}


async def test_set(db):
    with open(db.filename) as f:
        assert f.readline() == 'greeting,{"hello": "world"}\n'
        assert f.readline() == 'menu,{"breakfast": "bubur ayam", "lunch": "nasi rendang", "dinner": "nasi goreng"}\n'


async def test_get(db):
    assert await db.get('menu') == {"breakfast": "bubur ayam", "lunch": "nasi rendang", "dinner": "nasi goreng"}
    assert await db.get('greeting') == {"hello": "world"}


async def test_get_invalid_key(db):
    assert await db.get('invalid key') == None

File: 0001_simple_db_in_memory_index/cpp/test_simple_db_in_memory_index.cpp
#include "simple_db_in_memory_index.h"
#include <cstdio>
#include <fstream>
#include <gtest/gtest.h>

class SimpleDbInMemoryIndexTest : public ::testing::Test {
protected:
  const std::string filename = "test.db";
  SimpleDbInMemoryIndex *db;

  const nlohmann::json greeting_json = {{"hello", "world"}};
  const nlohmann::json menu_json = {{"breakfast", "bubur ayam"},
                                    {"lunch", "nasi rendang"},
                                    {"dinner", "nasi goreng"}};

  void SetUp() override {
    db = new SimpleDbInMemoryIndex(filename);
    db->set("greeting", greeting_json);
    db->set("menu", menu_json);
  }

  void TearDown() override {
    delete db;
    std::remove(filename.c_str());
  }
};

// Test case for existing database
TEST_F(SimpleDbInMemoryIndexTest, ExistingDb) {
  SimpleDbInMemoryIndex db2(filename);
  auto idx_map =
      db2.get_index().get_idx_map();
  std::unordered_map<std::string, std::pair<long, size_t>> expected_idx_map = {
      {"greeting", {0, 27}}, {"menu", {27, 78}}};
  EXPECT_EQ(idx_map, expected_idx_map);
  EXPECT_EQ(db2.get("menu"), menu_json);
  EXPECT_EQ(db2.get("greeting"), greeting_json);
}

// Test case for setting values
TEST_F(SimpleDbInMemoryIndexTest, Set) {
  std::ifstream file(db->get_filename());
  ASSERT_TRUE(file.is_open());
  std::string line;
  std::getline(file, line);
  EXPECT_EQ(line, "greeting,{\"hello\":\"world\"}");
  std::getline(file, line);
  EXPECT_EQ(line, "menu,{\"breakfast\":\"bubur ayam\",\"dinner\":\"nasi "
                  "goreng\",\"lunch\":\"nasi rendang\"}");
  file.close();
}

// Test case for getting values
TEST_F(SimpleDbInMemoryIndexTest, Get) {
  EXPECT_EQ(db->get("menu"), menu_json);
  EXPECT_EQ(db->get("greeting"), greeting_json);
}

// Test case for getting invalid key
TEST_F(SimpleDbInMemoryIndexTest, GetInvalidKey) {
  EXPECT_EQ(db->get("invalid key"), nullptr);
}

int main(int argc, char **argv) {
  ::testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}